source/common/ssl/context_config_impl.cc

```diff
@@ -138,9 +132,26 @@ unsigned ContextConfigImpl::tlsVersionFromProto(
   NOT_REACHED;
 }
 
+const std::string& ContextConfigImpl::certChain() const {
```

Comment:
If init fails and we are left with no credential, what is the exact end to end flow? Where will things fail and how will they fail (in both the listener and cluster case)? Is that handled in this PR or a follow up? Are you planning on adding an integration test for SDS? Can this integration test be done in this PR?

Reply:
Hi @mattklein123 This is the design for failure handling.  If SDS fails to get required secrets, either grpc connection fails, or response is empty or invalid response,  initialization will not continue.  The whole object, ListenerImpl, or ClusterImpl will stuck in Initialization phase and remains in the warning list.  Will this work?

Reply:
> The whole object, ListenerImpl, or ClusterImpl will stuck in Initialization phase and remains in the warning list.

No, this is not OK. Nothing can get stuck in the warming list. This is how it needs to work:
1) We attempt to use init manager for warming SDS for clusters/listeners.
2) If this succeeds, we have a credential and move forward.
3) If it fails, we have no credential and move forward.
4) If there is no credential, for listeners, we need to increment some failure stat and immediately close connections when they come in.
5) If there is no credential, for clusters, we need to increment some failure stat and immediately fail requests/connections to the upstreams.
6) If a credential is eventually pushed to the listener/cluster, things would start working.

I would like to see integration tests for all of these scenarios to cover the full SDS flow. Thanks!

Reply:
> Nothing can get stuck in the warming list.

@mattklein123 I have updated sds_api.cc, HostImpl::createConnection() in upstream_impl.cc, and  ConnectionHandlerImpl::ActiveListener::newConnection() to avoid getting stuck in the warming list.

If SDS does not have secrets when Envoy starts, the ContextConfigImpl::isValid() returns false. ContextManagerImpl::createSslClientContext() and ContextManagerImpl::createSslServerContext() check isValid(), and return nullptr as client/server context in the case that isValid() returns false. 

Accordingly, ConnectionHandlerImpl::ActiveListener::newConnection() updates stats downstream_cx_destroy_by_sds, and closes incoming downstream connection. HostImpl::createConnection() updates stats upstream_cx_connect_fail_by_sds and make connection creation fail. When SDS gets secrets, then isValid() returns true, and client/server context could be created and new connections are handled successfully. 

We are still working on adding integration tests, at meantime, I would like to get some feedbacks on the implementation. Thanks!

Reply:
Hi @mattklein123 in order NOT to make this already big PR too big,  we decided to add integration tests in the next PR. 
This PR is ready for review.  Thanks.

Reply:
Thanks you @qiwzhang I think this will be great for quality. In this PR I would do the following integration tests:
1) SDS listener config, SDS fails. Make sure connections fail and inc stat.
2) SDS listener config, SDS succeeds on initial warm. Make sure things work.
3) SDS cluster config, SDS fails. Make sure upstream connections fail/503 and inc stat.
4) SDS cluster config, SDS succeeds on initial warm. Make sure things work.

Once this is in place, adding additional integration tests for the update flow in the next PR should be pretty "easy."